cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -237,6 +279,7 @@ protected void run() throws Exception {
     final Iterator<WorkflowNode> iterator = workflowSpec.getNodes().iterator();
     while (iterator.hasNext() && runningThread != null) {
       try {
+        checkSuspended();
```

Comment:
Looks like code duplication to me. You better have a method `executeAll(Iterator<WorkflowNode> iterator)` to encapsulate the logic.